drivers/ethernet/phy/phy_microchip_t1s.c

```diff
@@ -0,0 +1,561 @@
+				   LAN865X_CFGPARAM_READ_ENABLE);
+	if (ret) {
+		return ret;
+	}
+
+	ret = phy_mc_t1s_c45_read(dev, MDIO_MMD_VENDOR_SPECIFIC2, LAN865X_REG_CFGPARAM_DATA, &val);
+	if (ret) {
+		return ret;
+	}
+
+	return val;
+}
+
+static int lan865x_generate_cfg_offsets(const struct device *dev, int8_t offsets[])
+{
+	const uint16_t fixup_regs[2] = {0x0004, 0x0008};
+	int ret;
+
+	for (int i = 0; i < ARRAY_SIZE(fixup_regs); i++) {
+		ret = lan865x_indirect_read(dev, fixup_regs[i]);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* 5-bit signed value, sign extend */
+		ret &= GENMASK(4, 0);
+		if (ret & BIT(4)) {
+			offsets[i] = ret | 0xE0;
+		} else {
+			offsets[i] = ret;
+		}
+	}
+
+	return 0;
+}
+
+static int lan865x_read_cfg_params(const struct device *dev, const uint16_t cfg_regs[],
+				   uint16_t cfg_params[], uint8_t count)
+{
+	uint16_t val;
+	int ret;
+
+	for (int i = 0; i < count; i++) {
+		ret = phy_mc_t1s_c45_read(dev, MDIO_MMD_VENDOR_SPECIFIC2, cfg_regs[i], &val);
+		if (ret < 0) {
+			return ret;
+		}
+		cfg_params[i] = (uint16_t)val;
+	}
+
+	return 0;
+}
+
+static int lan865x_write_cfg_params(const struct device *dev, const uint16_t cfg_regs[],
+				    uint16_t cfg_params[], uint8_t count)
+{
+	int ret;
+
+	for (int i = 0; i < count; i++) {
+		ret = phy_mc_t1s_c45_write(dev, MDIO_MMD_VENDOR_SPECIFIC2, cfg_regs[i],
+					   cfg_params[i]);
+		if (ret) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int lan865x_setup_cfgparam(const struct device *dev, int8_t offsets[])
+{
+	uint16_t cfg_results[ARRAY_SIZE(lan865x_revb_fixup_cfg_regs)];
+	uint16_t cfg_params[ARRAY_SIZE(lan865x_revb_fixup_cfg_regs)];
+	int ret;
+
+	ret = lan865x_read_cfg_params(dev, lan865x_revb_fixup_cfg_regs, cfg_params,
+				      ARRAY_SIZE(cfg_params));
+	if (ret) {
+		return ret;
+	}
+
+	cfg_results[0] = FIELD_PREP(GENMASK(15, 10), 9 + offsets[0]) |
+			 FIELD_PREP(GENMASK(9, 4), 14 + offsets[0]) | 0x03;
+	cfg_results[1] = FIELD_PREP(GENMASK(15, 10), 40 + offsets[1]);
+
+	return lan865x_write_cfg_params(dev, lan865x_revb_fixup_cfg_regs, cfg_results,
+					ARRAY_SIZE(cfg_results));
+}
+
+static int lan865x_setup_sqi_cfgparam(const struct device *dev, int8_t offsets[])
+{
+	uint16_t cfg_results[ARRAY_SIZE(lan865x_revb_sqi_fixup_cfg_regs)];
+	uint16_t cfg_params[ARRAY_SIZE(lan865x_revb_sqi_fixup_cfg_regs)];
+	int ret;
+
+	ret = lan865x_read_cfg_params(dev, lan865x_revb_sqi_fixup_cfg_regs, cfg_params,
+				      ARRAY_SIZE(cfg_params));
+	if (ret) {
+		return ret;
+	}
+
+	cfg_results[0] = FIELD_PREP(GENMASK(13, 8), 5 + offsets[0]) | (9 + offsets[0]);
+	cfg_results[1] = FIELD_PREP(GENMASK(13, 8), 9 + offsets[0]) | (14 + offsets[0]);
+	cfg_results[2] = FIELD_PREP(GENMASK(13, 8), 17 + offsets[0]) | (22 + offsets[0]);
+
+	return lan865x_write_cfg_params(dev, lan865x_revb_sqi_fixup_cfg_regs, cfg_results,
+					ARRAY_SIZE(cfg_results));
+}
+
+static int phy_mc_lan865x_revb_config_init(const struct device *dev)
+{
+	int8_t offsets[2];
```

Comment:
Isn't this configuration similar to what has been added to:
https://github.com/zephyrproject-rtos/zephyr/pull/80639

Reply:
Many parts of this file re-implements features from:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_lan865x.c

Either we move them to PHY (in a generic way) or we reuse those from eth_lan865x.c so this patch would be smaller.

Reply:
Hi,

Yes, you are right. By consolidating all your comments, please refer the below clarification. The below changes are needed to provide LAN8670/1/2 PHY driver support. The basic idea for this PR is to remove the code duplication when further T1S PHY drivers are getting in.

The idea is to separate MAC, OA TC6 Lib, OA TC14 PLCA Lib and T1S PHY drivers. As of now I have done the below implementations,

1. Implemented Microchip's T1S PHYs driver which includes all the supported T1S PHYs and latest revisions.
    drivers/ethernet/phy/phy_microchip_t1s.c
2. Implemented OA TC14 PLCA generic lib which can be used by all the 10Base-T1S PHYs to configure PLCA settings so that all the PHY drivers need not to duplicate the same code for that.
   drivers/ethernet/phy/phy_oa_tc14_plca.c
3. Implemented PHY C45 direct access support in the PHY interface module so that the OA TC6 specified MAC-PHYs can use this feature to write C45 registers directly instead of C22 indirect approach.
   include/zephyr/net/phy.h

Hope this clarifies the purpose of this PR. Please let me know if you need anything further. Kindly comment if you have anything in the implemented code to get this PR in.

Currently I am preparing the eth_lan865x.c and oa_tc6.c according to the above requirements like removing the duplicated code as you pointed out and adding the needed APIs or interface. Soon I will send one more PR with these changes so that everything will be aligned properly.

Reply:
@ParthibanI17164 - Thanks for the explanation.

In principle I do agree with your idea to have the different functionalities moved to separate files (like PHY, PLCA setup).
This would clarify things and help with maintenance.

> Currently I am preparing the eth_lan865x.c and oa_tc6.c according to the above requirements like removing the duplicated code as you pointed out and adding the needed APIs or interface. Soon I will send one more PR with these changes so that everything will be aligned properly.

Yes, this would help, as I could test your changes in a real-life setup (not only with code review).

@ParthibanI17164 - do you plan to update commits from this PR or shall I review those already provided?

Reply:
> @ParthibanI17164 - Thanks for the explanation.
> 
> In principle I do agree with your idea to have the different functionalities moved to separate files (like PHY, PLCA setup). This would clarify things and help with maintenance.
@lmajewski : Thanks for your feedback.
> 
> > Currently I am preparing the eth_lan865x.c and oa_tc6.c according to the above requirements like removing the duplicated code as you pointed out and adding the needed APIs or interface. Soon I will send one more PR with these changes so that everything will be aligned properly.
> 
> Yes, this would help, as I could test your changes in a real-life setup (not only with code review).
@lmajewski : That will be really helpful. Thank you very much for your support.
> 
> @ParthibanI17164 - do you plan to update commits from this PR or shall I review those already provided?
@lmajewski : Yes this was the question from my side as well. I would prefer to have those commits in this PR itself so that it will be easy for you to review and test instead of creating a new PR. It you agree with this proposal, shall I update this PR with those new commits and the comments you provided so far and then can you continue your review and testing? In this case, I will try to include them as soon as possible and let you know to continue again.

Reply:
Great. Thanks for your confirmation. I will add the commits in this PR and let you know.